core/src/main/scala/io/qbeast/spark/index/SparkPlanAnalyzer.scala

```diff
@@ -0,0 +1,120 @@
+package io.qbeast.spark.index
+
+import io.qbeast.spark.internal.rules.QbeastRelation
+import org.apache.spark.sql.catalyst.expressions.Alias
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.plans.logical.Filter
+import org.apache.spark.sql.catalyst.plans.logical.LocalLimit
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.catalyst.plans.logical.Sample
+import org.apache.spark.sql.catalyst.plans.logical.Sort
+import org.apache.spark.sql.DataFrame
+
+import scala.collection.convert.ImplicitConversions.`collection asJava`
+
+trait SparkPlanAnalyzer {
+
+  /**
+   * Builds a list of non-deterministic operations in the logical plan
+   *
+   * A list of Non-Deterministic operations:
+   *   - LocalLimit without Sort Operation
+   *   - Sample
+   *   - Filter with non-deterministic condition
+   *
+   * @param plan
+   *   the logical plan
+   * @return
+   */
+  private def isLogicalPlanDeterministic(plan: LogicalPlan): Boolean = {
+    // Recursively traverse the logical plan to find non-deterministic operations
+    val isCurrentOperationDeterministic = plan match {
+      case LocalLimit(_, _: Sort) => true // LocalLimit with Sort is deterministic
+      case LocalLimit(_, _) => false
+      case Sample(_, _, false, _, child) =>
+        child match {
+          case QbeastRelation(_, _) =>
+            true // Sample over QbeastRelation is deterministic
+          case _ => false
+        }
+      case Filter(condition, _) => condition.deterministic
+      case _ => true
+    }
+
+    val areChildOperationsDeterministic = plan.children.forall(isLogicalPlanDeterministic)
+
+    isCurrentOperationDeterministic && areChildOperationsDeterministic
+  }
+
+  /**
+   * Extracts the select expressions from the logical plan for the provided column name
+   *
+   * @param logicalPlan
+   *   the logical plan
+   * @param columnName
+   *   the column name
+   * @return
+   */
+  private[index] def collectSelectExpressions(
```

Comment:
Does this catch the case where an indexing column `b` is built using a non-deterministic column `a` and `a` is not indexed?

Reply:
It should, but let me double chect it